### PR TITLE
implement a lazy evaluation / thunk compiler primitive

### PR DESCRIPTION
**Pull request by DanilaFe**
# The Problem

I would like to be able to write module code for "built-in" Chapel features. That's because some code is much easier to express in plain Chapel than in AST manipulation. For instance, I would like to be able to use `on` clauses and `forall` loops to implement language features, rather than having to build lowered forms of them in the AST. It's also significantly easier to change the Chapel module implementation than having to modify the AST-building compiler code. For instance, consider the following code that I would like to use to implement remote variable declarations:

```Chapel
    var c: owned _remoteVarContainer(inType)?;
    on loc do c = new _remoteVarContainer(expr);
    return new _remoteVarWrapper(try! c : owned _remoteVarContainer(inType));
```

There are a lot of AST nodes here, and this is just one of the three versions of this code (it handles the case where both the type and the value of the remote variable are specified). Writing it as C++ `buildWhateverStatement` is error prone.

And yet, __I can't just turn that code into a Chapel function and call it__:

```Chapel
proc myFn(loc: locale, type inType, value: inType) {
    var c: owned _remoteVarContainer(inType)?;
    on loc do c = new _remoteVarContainer(value);
    return new _remoteVarWrapper(try! c : owned _remoteVarContainer(inType));
}

var myWrapper = myFn(loc, inType, expr); // not the same!
```

The reason is that arguments to functions are evaluated before they are given to function bodies. In my example, `expr` will get evaluated before the call to `myFn`, and thus the computation will not occur on `loc`. This is currently causing a problem for remote variables (my second snippet matches my actual implementation). Static variables suffer from this problem because the compiler-generated C++ code is brittle and long.

# The Solution

What I'd like to be able to do is to pass in an "expression" into Chapel code, to be evaluated by the function when needed, be it on a different locale or conditionally. To this end, what I need is to be able to defer computations, and that's what this PR adds: [thunks](https://en.wikipedia.org/wiki/Thunk). Then, I can write code like this:

```Chapel
proc myFn(loc: locale, type inType, in thunk: _thunkRecord) {
    var c: owned _remoteVarContainer(inType)?;
    on loc do c = new _remoteVarContainer(__primitive("force thunk", thunk));
    return new _remoteVarWrapper(try! c : owned _remoteVarContainer(inType));
}

var myWrapper = myFn(loc, inType, __primitive("create thunk", expr)); // same as the "inlined" form!
```

Another example (from a test file) is the following:

```Chapel
proc executeIfTrue(cond: bool, in thunk: _thunkRecord) {
  var temp: thunkToReturnType(thunk.type);
  if cond {
    temp = __primitive("force thunk", thunk);
  }
  return temp;
}

writeln(executeIfTrue(true, __primitive("create thunk", new C?(42)))); // prints "calling C.init", then "{42}"
writeln(executeIfTrue(false, __primitive("create thunk", new C?(42)))); // doesn't print, then prints "nil"
```

# Implementation
This works by doing pretty much what we do for iterators and their records/classes: for each thunked expression (the `create thunk` primitive), it creates a new builder function `chpl__thunkN` and, eventually, a record `_tr_chpl__thunkN`. It re-uses the same logic for capturing outer variables etc, so that the expression being captured can refer to global or local variables, fields, etc. In detail, the process is as follows:

1. __Pass 8 - Normalize__: lifts `create thunk` primitives into builder functions, named `chpl__thunkN`. These functions accept all the captured variables as formals. Eventually, these functions are marked to return the thunk record; however, during normalize, the thunk record doesn't yet exist. The thunk body -- the expression being deferred -- is copied into these functions.
2. __Pass 12 - Resolve__: after resolving the types of the `chpl__thunkN` function, the function resolution code creates the thunk record (but doesn't yet populate its fields; this matches how iterator records are handled). It also creates the `invoke` method, which is called by `force thunk` primitives. The 'invoke' method is empty for the time being (again, like the prototype methods for iterators) and is populated once the thunk record's fields have been created. The `force thunk` primitive is simply rewritten to the `invoke` method.
3. __Pass 20 - Lower Iterators__: at this time, the thunk record is populated with the fields that correspond to the captured variables, and the body of the `invoke` method is filled in with the original code that the user provided; references to outer variables are replaced with field references to the thunk record. The builder function is made to create a new thunk record and populate its fields with the formals, preparing it for invocation.

Much of the logic is shared with lowering iterators, and I have factored out code where appropriate to ensure that little-to-no code duplication is involved.

# Next steps
Since this PR is relatively large, I didn't want to also include into it changes to remote or static varibles. However, I believe that the next step is to switch the remote variable implementation to use thunks, which will allow it to consistently execute the initialization expression on the target locale (https://github.com/chapel-lang/chapel/issues/25298).

Reviewed by @e-kayrakli -- thanks!

# Testing
- [x] paratest
- [x] paratest gasnet
- [x] GPU test

### CODE REVIEW

File: compiler/passes/normalize.cpp [hunk 1]
```diff
@@ -694,6 +695,88 @@ static void insertCallTempsForRiSpecs(BaseAST* base) {
...
+static CallExpr* buildThunkPrimFunctions(CallExpr* node) {
+  CHPL_ASSERT(node->numActuals() == 1);
+  auto delayedExpr = node->get(1);
+
+  std::set<Symbol*> outerVars;
+  collectThunkOuterVars(delayedExpr, outerVars);
+
+  static int thunkUid = 0;
+  thunkUid++;
+  auto thunkUidStr = istr(thunkUid);
```
Comment (e-kayrakli) [inline compiler/passes/normalize.cpp:727]: Why did you need this enumeration/mangling here?

File: compiler/resolution/resolveFunction.cpp [hunk 2]
```diff
@@ -516,6 +516,78 @@ void resolveSpecifiedReturnType(FnSymbol* fn) {
...
+static void setReturnAndReturnSymbolType(FnSymbol* fn, Type* retType) {
+  fn->retType = retType;
+  Symbol* retSym = fn->getReturnSymbol();
+  INT_ASSERT(retSym);
+  retSym->type = retType;
+  fn->retTag = RET_VALUE;
+}
+
+static FnSymbol* buildThunkInvokeMethod(FnSymbol* fn, Type* thunkResultType, AggregateType* newRecord) {
+
```
Comment (e-kayrakli) [inline compiler/resolution/resolveFunction.cpp:550]: drop empty line

File: compiler/resolution/resolveFunction.cpp [hunk 3]
```diff
@@ -516,6 +516,78 @@ void resolveSpecifiedReturnType(FnSymbol* fn) {
...
+static void setReturnAndReturnSymbolType(FnSymbol* fn, Type* retType) {
+  fn->retType = retType;
+  Symbol* retSym = fn->getReturnSymbol();
+  INT_ASSERT(retSym);
+  retSym->type = retType;
+  fn->retTag = RET_VALUE;
+}
+
+static FnSymbol* buildThunkInvokeMethod(FnSymbol* fn, Type* thunkResultType, AggregateType* newRecord) {
```
Comment (e-kayrakli) [inline compiler/resolution/resolveFunction.cpp:549]: reflow to 80 characters

File: compiler/resolution/resolveFunction.cpp [hunk 4]
```diff
@@ -516,6 +516,78 @@ void resolveSpecifiedReturnType(FnSymbol* fn) {
...
+  auto invokeFn = buildThunkInvokeMethod(fn, thunkResultType, newRecord);
+
+  // Adjust the function's return type and the type of the return symbol
+  setReturnAndReturnSymbolType(fn, newRecord);
+
+  fn->defPoint->insertBefore(new DefExpr(newRecord->symbol));
+  fn->defPoint->insertBefore(new DefExpr(invokeFn));
+
+  makeRefType(newRecord);
+  normalize(invokeFn);
```
Comment (e-kayrakli) [inline compiler/resolution/resolveFunction.cpp:584]: Do you know what it would take to not call `normalize` here and create the AST in a normalized way? Whenever I call `normalize` later than the pass `normalize` things seem to get messed up.

File: test/variables/thunks/thunk-arrays.chpl [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+proc doSomething(r: R) {
+  writeln("doing something else!");
+  return 42 + r.x;
+}
+
+proc doSomething(c: C) { // Yes, generic.
+  writeln("doing something else!");
+  return 42 + c.x;
+}
```
Comment (e-kayrakli) [inline test/variables/thunks/thunk-arrays.chpl:17]: Leftover from other code?

**Review by e-kayrakli — COMMENTED** [review]
Nice work! I have some stylistic preferences and suggestions that you may or may not take.

My questions on a test's behavior, suggestions to use a more object oriented approach in the compiler and creating a file for thunk lowering are on the more serious side. Please let me know what you think of those.

File: compiler/AST/iterator.cpp [hunk 6]
```diff
@@ -425,11 +425,6 @@ removeRetSymbolAndUses(FnSymbol* fn) {
 
   // We cannot remove rsym's definition, because rsym
   // may also be referenced in an autoDestroy call.
-
-  INT_ASSERT(fn->iteratorInfo != NULL);
-  Type* yieldedType = fn->iteratorInfo->yieldedType;
-
-  return yieldedType;
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/iterator.cpp:432]: I inlined this part of the code into the function that uses it. It doesn't fit the function name and it doesn't relate to the function's other operations. In addition, iterator records needed this line, but not thunks. So I inlined it in the one place where it's used.

File: compiler/AST/iterator.cpp [hunk 7]
```diff
@@ -862,6 +857,39 @@ static void replaceLocalWithFieldTemp(SymExpr*       se,
...
+
+    if (call && call->isPrimitive(PRIM_ADDR_OF)) {
+
+      // Convert (addr of var) to (. _ic field).
+      // Note, GET_MEMBER is not valid on a ref field;
+      // in that event, GET_MEMBER_VALUE returns the ref.
+      if (field->isRef())
+        call->primitive = primitives[PRIM_GET_MEMBER_VALUE];
+      else
+        call->primitive = primitives[PRIM_GET_MEMBER];
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/iterator.cpp:883]: I think when debugging primitives -- and how they pop in and out of existence -- I actually grep the source code for the primitive name, and thus would fine a `new CallExpr` and `primitives[...]` just the same. That said, it's definitely magical when an AST node changes primitive in the middle of nowhere, and it would get in the way of the "traditional" mode of Chapel debugging.

File: compiler/AST/iterator.cpp [hunk 8]
```diff
@@ -2128,11 +2149,139 @@ static void addLocalsToClassAndRecord(Vec<Symbol*>& locals, FnSymbol* fn,
...
+    for_formals(formal, fn) {
+      if (formal->hasFlag(FLAG_RETARG)) returnType = formal->type->getValType();
+    }
+  }
+
+  CHPL_ASSERT(returnType);
+  auto recordType = toAggregateType(returnType);
+  CHPL_ASSERT(recordType);
+  CHPL_ASSERT(recordType->aggregateTag == AGGREGATE_RECORD);
+  CHPL_ASSERT(recordType->symbol->hasFlag(FLAG_THUNK_RECORD));
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/iterator.cpp:2171]: It's a dyno-ism, but I intended for `INT_ASSERT`.

File: compiler/AST/iterator.cpp [hunk 9]
```diff
@@ -2128,11 +2149,139 @@ static void addLocalsToClassAndRecord(Vec<Symbol*>& locals, FnSymbol* fn,
...
+}
+
+static void populateThunkInvokeFn(FnSymbol* fn, FnSymbol* invokeFn) {
+  auto thunkBody = fn->body->body.last();
+  invokeFn->body->replace(thunkBody->remove());
+
+  // Find the thunk result, transform it into a return at the end of the function.
+  for_alist_backward(expr, invokeFn->body->body) {
+    if (auto call = toCallExpr(expr)) {
+      if (call->isPrimitive(PRIM_THUNK_RESULT)) {
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/iterator.cpp:2212]: > The reason I think of that is here, you're just finding a call that you've created couple of functions ago.

I don't think it's quite so simple. the `PRIM_THUNK_RESULT` is created at pass 12, whereas thunk lowering happens at pass 20. I could store some kind of global map of thunk builders to their results, but it seems easier to just find the primitive in the tree (besides, what if there are instantiations etc.?)

File: compiler/passes/checkResolved.cpp [hunk 10]
```diff
@@ -541,13 +541,15 @@ static void
 checkReturnPaths(FnSymbol* fn) {
   // Check to see if the function returns a value.
   if (fn->isIterator() ||
+      fn->hasFlag(FLAG_THUNK_BUILDER) ||
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/passes/checkResolved.cpp:544]: Surprisingly, yes -- `isIterator` is the iterator record analog to `FLAG_THUNK_BUILDER` (function that returns an iterator / function that returns a thunk record) whereas `FLAG_AUTO_II` is the iterator analog to `FLAG_THUNK_INVOKE` (methods like `zip` etc / method `invoke`). So I put them near their analogs.

File: compiler/passes/normalize.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/passes/normalize.cpp:727]: I don't know that I strictly need it but it matches what we do with loop expressions etc. (each loop expression gets its own number).

File: test/variables/thunks/thunk-arrays.chpl (hunk 5 above)
Comment (DanilaFe) [reply test/variables/thunks/thunk-arrays.chpl:17]: Yup.

File: test/variables/thunks/thunk-capture.chpl [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  writeln("doing something!");
+  return 42 + input;
+}
+
+proc doSomething(r: R) {
+  writeln("doing something else!");
+  return 42 + r.x;
+}
+
+proc doSomething(c: C) { // Yes, generic.
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply test/variables/thunks/thunk-capture.chpl:14]: Yes, it's a generic function because of ownership. I'll add a clarification.

File: test/variables/thunks/thunk-cond.chpl [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  var temp: thunkToReturnType(thunk.type);
+  writeln("In execute if true, with value: ", cond);
+  if cond {
+    temp = __primitive("force thunk", thunk);
+  }
+  return temp;
+}
+
+writeln(executeIfTrue(true, __primitive("create thunk", new C?(42))));
+writeln(executeIfTrue(false, __primitive("create thunk", new C?(42))));
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply test/variables/thunks/thunk-cond.chpl:21]: Yep, `execIfTrue` is a generic function; if it were instantiated with a thunk that produces a non-nilable type, this line would fail to compile:

```Chapel
  var temp: thunkToReturnType(thunk.type);
```

Since it would be

```
  var temp: owned C;
```

Which doesn't have a viable default-initialization.

File: compiler/AST/LoopExpr.cpp [hunk 13]
```diff
@@ -623,7 +623,7 @@ static void findOuterVars(LoopExpr* loopExpr,
   }
 }
 
-static ArgSymbol* newOuterVarArg(Symbol* ovar) {
+ArgSymbol* newOuterVarArg(Symbol* ovar) {
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/LoopExpr.cpp:626]: I added comments because I can't think of a more descriptive name for `newOuterVarArg` that isn't too long.

File: compiler/AST/iterator.cpp [hunk 14]
```diff
@@ -2128,11 +2149,139 @@ static void addLocalsToClassAndRecord(Vec<Symbol*>& locals, FnSymbol* fn,
...
+
+  initializeRecordFieldWithArgLocals(fn, thunk, locals, local2field);
+
+  // Return the filled-in iterator record.
+  insertReturn(fn, thunk);
+
+  fn->addFlag(FLAG_INLINE);
+}
+
+void lowerThunk(FnSymbol* fn) {
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/AST/iterator.cpp:2247]: I did this. It required making some static functions in `iterator.h` public, since a lot of the logic from `iterator.cpp` is shared. Let me know if you think that's fine.

File: compiler/include/iterator.h [hunk 15]
```diff
@@ -85,6 +85,7 @@ void gatherPrimIRFieldValByFormal();
 void cleanupPrimIRFieldValByFormal();
 
 void lowerIterator(FnSymbol* fn);
+void lowerThunk(FnSymbol* fn);
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/include/iterator.h:88]: To mee, `resolution.h` is a misleading place since this occurs in pass 20, lowerIterators. I think it makes sense to leve `lowerThunk` near `lowerIterator` since they occur in the same pass and behave very similarly. I'm also open to other places -- or other interpretations of "resolution.h" :)

File: compiler/resolution/resolveFunction.cpp (hunk 4 above)
Comment (DanilaFe) [reply compiler/resolution/resolveFunction.cpp:584]: The existing iterator support (see `protoIteratorClass`) seems to call `normalize` and not have any issues. I could try writing normalized code, but it seems the precedent says it's fine to do this.

File: compiler/AST/iterator.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/AST/iterator.cpp:877]: Yeah. I don't think we do this particular modification you're doing there too much. But I think this is a case where `CallExpr::setPrimitive` with a hook for `break-on-something-id` would be interesting if this pattern was more prevalent.

Your workflow of grepping is something I do as well, but I don't think that would work efficiently for the primitives you have here as I am sure we have a ton of instances of those.

In any case, as I said before, I am not asking for change here. I don't think this is ideal, but you're just moving the code and reaching my ideal is non-trivial for this PR.

Resolving this conversation.

File: compiler/AST/iterator.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/AST/iterator.cpp]: I think that's a step in the right direction certainly. A more ideal solution (that I am not asking for) is to put those functions in a common header (I am not sure what I would name that header myself) to be shared between the iterator and the thunk logic.

No further action needed for my sake, though.

File: compiler/AST/iterator.cpp [hunk 16]
```diff
@@ -2128,11 +2149,139 @@ static void addLocalsToClassAndRecord(Vec<Symbol*>& locals, FnSymbol* fn,
...
 }
 
+static AggregateType* getThunkBuilderReturnType(FnSymbol* fn) {
+  Type* returnType = nullptr;
+  if (!fn->hasFlag(FLAG_FN_RETARG)) {
+    returnType = fn->retType->getValType();
+  } else {
+    for_formals(formal, fn) {
+      if (formal->hasFlag(FLAG_RETARG)) returnType = formal->type->getValType();
+    }
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/AST/iterator.cpp]: Feel free to resolve this if you're not willing to add that helper.

File: compiler/AST/iterator.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/AST/iterator.cpp]: Oh, OK, yeah I am all too familiar with that problem. It wasn't clear to me reading the code linearly. Could you add some comments in the new implementation file you created for these functions where you have sections for parts that execute during normalization and then pass 20 (callDestructors?)?

File: compiler/include/LoopExpr.h [hunk 17]
```diff
@@ -70,6 +70,8 @@ class LoopExpr final : public Expr {
 };
 
 bool considerForOuter(Symbol* sym);
+ArgSymbol* newOuterVarArg(Symbol* sym);
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/include/LoopExpr.h:96]: You added a comment. Resolving.

File: compiler/include/iterator.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/include/iterator.h:130]: The more I see problems like this, the more I want to have `thunk.h`. Or if I was the king of the world, I would just make this logic its own pass.

I am not proud of the pattern in our compiler code where we stick things related or unrelated next to things that happen right before or after them. That's another thing that I've been doing but trying to steer away from as much as I can these days. This particular case looks like it could be easily handled in a way to avoid that. If you agree with that, that is.

I don't intend to try to convince you. My current preference is to have a `thunk.h`. But I can go with what your final call is.

File: compiler/passes/checkResolved.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/passes/checkResolved.cpp:544]: I see. Could you put comments next to them pointing those similarities out? This could strike someone as odd, and comments could help understand things a bit better.

File: compiler/passes/normalize.cpp (hunk 1 above)
Comment (e-kayrakli) [reply compiler/passes/normalize.cpp:727]: I see. OK, yeah. Admittedly, I relied on those enumerations in the logs before (maybe IDs could have sufficed, though). In any case, let's keep it.

Resolving.

File: compiler/resolution/resolveFunction.cpp (hunk 4 above)
Comment (e-kayrakli) [reply compiler/resolution/resolveFunction.cpp:587]: I am not crazy about the precedence, but trying to write the code in a normalized way is not always easy. Not something to bother with at this point.

Resolving.

File: test/variables/thunks/thunk-copies.good [hunk 18]
```diff
@@ -0,0 +1,9 @@
+(x = 42)
+deinitializing R with value 42
+(x = 43)
+deinitializing R with value 43
+deinitializing R with value 43
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/variables/thunks/thunk-copies.good:5]: OK, yeah, that makes sense. This made me think what happens for a shared class, but you've got that covered and single deinit for that case makes perfect sense.

File: compiler/AST/primitive.cpp [hunk 19]
```diff
@@ -730,6 +741,11 @@ initPrimitive() {
   prim_def(PRIM_OUTER_CONTEXT, "outer context", returnInfoFirst);
   prim_def(PRIM_HOIST_TO_CONTEXT, "hoist to context", returnInfoVoid);
 
+  prim_def(PRIM_CREATE_THUNK, "create thunk", returnInfoUnknown);
+  prim_def(PRIM_THUNK_RESULT, "thunk result", returnInfoFirst);
+  prim_def(PRIM_FORCE_THUNK, "force thunk", returnInfoUnknown);
+  prim_def(PRIM_THUNK_RESULT_TYPE, "thunk result type", returnInfoThunkResultType);
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/AST/primitive.cpp:747]: I am not sure if you have seen this comment in your first run.

**Review by e-kayrakli — APPROVED** [review]

File: compiler/include/iterator.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/include/iterator.h:130]: I've added a `thunks.h` and moved all implementation code from all passes into `thunks.cpp`. I also took that opportunity to add pass separators to `thunks.cpp`. I think it came out rather well.

[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/include/iterator.h]: Are we past the point of not adding new passes because of the compiler performance tracking challenges? Probably not. I am sorry if I led you down that path.

[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/include/iterator.h]: I didn't make a new pass (in fact, I don't think that I can, but only due to pragmatic concerns). I haven't thought about the general possibility.

[the remark this answers is not in the document]
Comment (e-kayrakli) [reply compiler/include/iterator.h]: Oh, OK. I misunderstood, then. Thanks for those changes.
